File: javascript/cormen/mergesort.js
var a = [8, 5, 3, 6, 4, 2, 9, 7, 1];

function merge(a, low, mid, high) {
    var i = low, j = mid + 1, k = low, b = [];
    while(i <= mid && j <= high) {
        if (a[i] < a[j]) {
            b.push(a[i++]);
        } else {
            b.push(a[j++]);
        }
    }
    while (i <= mid) b.push(a[i++]);
    while (j <= high) b.push(a[j++]);
    k = 0;
    for (var i = low; i <= high; i++) {
        a[i] = b[k++];
    }
}

function msort(a, low, high) {
    if (low < high) {
        var mid = Math.floor((low + high) / 2);
        msort(a, low, mid);
        msort(a, mid + 1, high);
        merge(a, low, mid, high);
    }
}

msort(a, 0, 8);
console.log(a);

File: cpp/problems/cormen/ch02/merge_sort.cpp
#include <iostream>

void merge(int a[], int low, int mid, int high) {
    int L[mid - low + 1];
    int R[high - mid + 1];
    int n1 = 0;
    int n2 = 0;
    for (int i = low; i <= mid; i++) {
        L[n1++] = a[i];
    }
    for (int i = mid + 1; i <= high; i++) {
        R[n2++] = a[i];
    }
    int i = 0, j = 0, k = low;
    while (i < n1 && j < n2) {
        if (L[i] < R[j]) a[k++] = L[i++];
        else a[k++] = R[j++];
    }
    while (i < n1) a[k++] = L[i++];
    while (j < n2) a[k++] = R[j++];
}

void merge_sort(int a[], int low, int high) {
    if (low < high) {
        int mid = (high + low) / 2;
        merge_sort(a, low, mid);
        merge_sort(a, mid + 1, high);
        merge(a, low, mid, high);
    }
}

int main(int argc, char* argv[]) {
    int a[] = {3, 6, 8, 2, 5, 9, 1, 4, 7};
    merge_sort(a, 0, 8);
    for (int i = 0; i < 9; i++) {
        std::cout << a[i] << " ";
    }
    return 0;
}
